math/src/field/fields/mersenne31/field.rs

```diff
@@ -0,0 +1,402 @@
+        let msb = n & (1 << 31);
+        let msb_reduced = msb >> 31;
+        let res = msb ^ n;
+
+        // assert msb_reduced fits within 31 bits
+        debug_assert!((res >> 31) == 0 && (msb_reduced >> 1) == 0);
+        res + msb_reduced
+    }
+
+    fn as_representative(n: &u32) -> u32 {
+        if *n == MERSENNE_31_PRIME_FIELD_ORDER {
+            0
+        } else {
+            *n
+        }
+    }
+}
+
+pub const MERSENNE_31_PRIME_FIELD_ORDER: u32 = (1 << 31) - 1;
+
+//NOTE: This implementation was inspired by and borrows from the work done by the Plonky3 team
+// https://github.com/Plonky3/Plonky3/blob/main/mersenne-31/src/lib.rs
+// Thank you for pushing this technology forward.
+impl IsField for Mersenne31Field {
+    type BaseType = u32;
+
+    /// Returns the sum of `a` and `b`.
+    fn add(a: &u32, b: &u32) -> u32 {
+        // Avoids conditional https://github.com/Plonky3/Plonky3/blob/6049a30c3b1f5351c3eb0f7c994dc97e8f68d10d/mersenne-31/src/lib.rs#L249
+        // Working with i32 means we get a flag which informs us if overflow happens
+        let (sum_i32, over) = (*a as i32).overflowing_add(*b as i32);
+        let sum_u32 = sum_i32 as u32;
+        let sum_corr = sum_u32.wrapping_sub(MERSENNE_31_PRIME_FIELD_ORDER);
+
+        //assert 31 bit clear
+        // If self + rhs did not overflow, return it.
+        // If self + rhs overflowed, sum_corr = self + rhs - (2**31 - 1).
+        let sum = if over { sum_corr } else { sum_u32 };
+        debug_assert!((sum >> 31) == 0);
+        Self::as_representative(&sum)
+    }
+
+    /// Returns the multiplication of `a` and `b`.
+    // Note: for powers of 2 we can perform bit shifting this would involve overriding the trait implementation
+    fn mul(a: &u32, b: &u32) -> u32 {
+        let prod = u64::from(*a) * u64::from(*b);
+        let prod_lo = (prod as u32) & ((1 << 31) - 1);
+        let prod_hi = (prod >> 31) as u32;
+        //assert prod_hi and prod_lo 31 bit clear
+        debug_assert!((prod_lo >> 31) == 0 && (prod_hi >> 31) == 0);
+        Self::add(&prod_lo, &prod_hi)
+    }
+
+    fn sub(a: &u32, b: &u32) -> u32 {
+        let (mut sub, over) = a.overflowing_sub(*b);
+
+        // If we didn't overflow we have the correct value.
+        // Otherwise we have added 2**32 = 2**31 + 1 mod 2**31 - 1.
+        // Hence we need to remove the most significant bit and subtract 1.
+        sub -= over as u32;
+        sub & MERSENNE_31_PRIME_FIELD_ORDER
+    }
+
+    /// Returns the additive inverse of `a`.
+    fn neg(a: &u32) -> u32 {
+        // NOTE: MODULUS known to have 31 bit clear
+        MERSENNE_31_PRIME_FIELD_ORDER - a
+    }
+
+    /// Returns the multiplicative inverse of `a`.
+    fn inv(a: &u32) -> Result<u32, FieldError> {
+        if *a == Self::zero() || *a == MERSENNE_31_PRIME_FIELD_ORDER {
+            return Err(FieldError::InvZeroError);
+        }
+        let p101 = Self::mul(&Self::pow(a, 4u32), a);
```

Comment:
Since we are doing powers of 2, it could be good at some time to implement a special version of pow to avoid the square and multiply steps